feat: add kotlin language support

IDontHaveBrain opened the pull request:
Add Kotlin Language Server Support

Document symbols
Go to definition
Find references
Hover information
Code completions

Tested on Windows-x64 platforms with the kotlin-language-server from the fwcd/kotlin-language-server repository.

LakshyAAAgrawal commented:
Hi @IDontHaveBrain

Thank you so much for the PRs you have kindly created.

At the moment the unit tests are failing. This is due to some external issue in the unit tests, which I have fixed in the latest commit to main branch.
Could you please merge/rebase on top of main branch to get the unit tests in the CI working again, after which I will review and merge the PR.

Could you please do this for all your branches?

IDontHaveBrain commented:
Thanks for the review.
1. Updated to fetch the Java environment directly in `setup_runtime_dependencies` method.
2. Refactored the error handling in the server shutdown sequence. Do you have any additional suggestions for improving this part?
Please review when you have time.

LakshyAAAgrawal commented on src/multilspy/lsp_protocol_handler/server.py:
```diff
@@ -234,27 +234,110 @@ async def stop(self) -> None:
-            # However, there's an issue with asyncio terminating processes documented at
-            # https://bugs.python.org/issue35539 and https://bugs.python.org/issue41320
-            # process.terminate()
-            wait_for_end = process.wait()
```
Why is this removed?

IDontHaveBrain replied:
Thank you for the review.
It still calls `asyncio.wait_for(process.wait(), timeout=10)`, just with a shorter timeout value.

_cleanup_process -> _terminate_or_kill_process
```
    async def _terminate_or_kill_process(self, process):
        """Try to terminate the process gracefully, then forcefully if necessary."""
        # First try to terminate the process tree gracefully
        self._signal_process_tree(process, terminate=True)

        # Wait for the process to exit (with timeout)
        try:
            await asyncio.wait_for(process.wait(), timeout=10)
```

LakshyAAAgrawal commented:
Hi @IDontHaveBrain, Thanks a lot for updating the PR. The changes looks great, can you just address the one comment I added above?

LakshyAAAgrawal commented on src/multilspy/language_servers/kotlin_language_server/kotlin_language_server.py:
```diff
@@ -0,0 +1,233 @@
+            await lsp.request_references(...)
+            # Shutdown the LanguageServer on exit from scope
+        # LanguageServer has been shutdown
+        """
```
nit: ``` mising

LakshyAAAgrawal commented on src/multilspy/language_servers/kotlin_language_server/runtime_dependencies.json:
```diff
@@ -0,0 +1,53 @@
+    },
+    "jdk": {
+        "win-x64": {
+            "url": "https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.6%2B7/OpenJDK21U-jdk_x64_windows_hotspot_21.0.6_7.zip",
```
Can you say if any of these can be replaced by the repos and urls at https://github.com/microsoft/multilspy/blob/127de628c87b26a5c3077a44e2b6ba80e23e5651/src/multilspy/language_servers/eclipse_jdtls/runtime_dependencies.json#L12

LakshyAAAgrawal reviewed (changes requested):
A few nits.

BTW, the PR is amazing, and I believe it solves some long-standing issues with multilspy. Thank you. Will merge after this final round of discussion.

LakshyAAAgrawal commented:
Also, add an entry to the README about Kotlin support in the table.

LakshyAAAgrawal commented on src/multilspy/language_servers/kotlin_language_server/runtime_dependencies.json:
```diff
@@ -0,0 +1,47 @@
+            "java_path": "extension/jre/17.0.8.1-win32-x86_64/bin/java.exe"
+        },
+        "win-arm64": {
+            "url": "https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.6%2B7/OpenJDK21U-jdk_aarch64_windows_hotspot_21.0.6_7.zip",
```
Is this not available through redhat?

IDontHaveBrain replied:
> Is this not available through redhat?

It seems that Red Hat OpenJDK does not support Windows ARM64 binaries.
At least, I couldn't find any in my search.

LakshyAAAgrawal replied:
https://blogs.windows.com/windows-insider/2020/12/10/introducing-x64-emulation-in-preview-for-windows-10-on-arm-pcs-to-the-windows-insider-program/

win32-x64 binaries can run on win-arm64 I believe. Could you just refer to that binary for the win-arm64 platform as well? Is it possible for you to test it? I don't have a win-arm machine.

IDontHaveBrain replied:
I understand that GitHub Actions supports Windows ARM64 environment.
will test it later.

LakshyAAAgrawal replied:
It's upto you: We can either test it before merging, or we can temporarily disable the win-arm64, merge it, track an issue to enable win-arm64 testing and add it in a separate PR.

IDontHaveBrain replied:
I've removed the win-arm64 section from the runtime_dependencies.json file.
It would be good to revisit this when we add cross-platform testing to our GitHub Actions workflows in the future.

Thank you for your kind review and assistance throughout this 
PR process!

LakshyAAAgrawal commented:
Thank you very much @IDontHaveBrain, this is a great PR!
